### PR TITLE
fix: fetch remainder of metadata if it is large

### LINKED ISSUE #856: Error: `panicked at 'assertion failed: file_size - manifest_pos <= buf.len()'`
I write a program to convert parquet files into lance format.
It converted 600+ parquet files successfully and then the error occurred.

```
panicked at 'assertion failed: file_size - manifest_pos <= buf.len()'
```

### PR DESCRIPTION
**Pull request by wjones127**
Fixes #856

### CODE REVIEW

**Comment by wjones127** [comment]
Fails with

```
called `Result::unwrap()` on an `Err` value: IO("failed to decode Protobuf message: Manifest.fields: invalid wire type: Varint (expected LengthDelimited)")
```

File: rust/src/io/reader.rs [hunk 1]
```diff
@@ -71,15 +75,34 @@ pub async fn read_manifest(object_store: &ObjectStore, path: &Path) -> Result<Ma
...
+            )
+            .await?
+            .into_iter()
+            .collect();
+        buf2.extend_from_slice(&buf);
+        buf2.freeze()
+    };
+
+    // Need to trim the magic number at end, and the non-manifest data at the beginning
+    let buf = buf.slice(4..buf.len() - 16);
```
Comment (eddyxu) [inline rust/src/io/reader.rs:103]: What does this `4` stand for?

File: rust/src/format/fragment.rs [hunk 2]
```diff
@@ -66,7 +66,7 @@ impl From<&pb::DataFile> for DataFile {
 ///
 /// A fragment is a set of files which represent the different columns of the same rows.
 /// If column exists in the schema, but the related file does not exist, treat this column as `nulls`.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq)]
```
Comment (eddyxu) [inline rust/src/format/fragment.rs:69]: Why does it require `PartialEq`, do you need to sort / storing them in some containers?

Comment (wjones127) [reply rust/src/format/fragment.rs:69]: This was for the sake of testing (`assert_eq` requires this).

File: rust/src/io/reader.rs (hunk 1 above)
Comment (wjones127) [reply rust/src/io/reader.rs:103]: hmm well my comment ", and the non-manifest data at the beginning" isn't right / specific. I'm not sure yet; all I knew is that it was skipped previously. I guess it's the u32 message length here:

https://github.com/eto-ai/lance/blob/d3f6f6c31c909da22c0b7d486bbc35ec101f6079/rust/src/io/object_writer.rs#L68

Perhaps we should be reading that length and validating it?

Comment (wjones127) [reply rust/src/io/reader.rs:104]: Added validation for that and a better comment 👍

**Review by eddyxu — APPROVED** [review]
LGTM

**Comment by haoxins** [comment]
can we bump a new version to crates.io?

**Comment by wjones127** [comment]
@haoxins we will release a new version today.

**Pull request merged**